flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java

```diff
@@ -265,7 +274,7 @@ protected void restoreJob(
             throws Exception {
         Optional<String> savepointOpt = Optional.empty();
 
-        if (spec.getJob().getUpgradeMode() != UpgradeMode.STATELESS) {
+        if (spec.getJob().getUpgradeMode() == UpgradeMode.SAVEPOINT) {
```

Comment:
nit: it seems appropriate to push this logic down into the deploy method. That would also eliminate usage of an optional as a method parameter (something that is debatably considered an antipattern and does not follow the original intention of the Optional type).

Reply:
The problem is that this cannot be easily done as what savepoint is passed there will depend on where / why the deploy is called. It is for example different during initial deployments vs upgrades.

While the optional parameter is not great I would like to avoid adding more complexity to the deploy method. It may make sense to actually break up the deploy method in 3 parts, stateless/last-state/savepoint which would get rid of the optional params + the requireHaMetadata flag at the same time at the expense of more methods.

Reply:
Whether this would actually makes sense from a code readability standpoint still remains to be seen